avalanche/training/templates/observation_type/batch_observation.py

```diff
@@ -22,36 +43,29 @@ def model_adaptation(self, model=None):
             # If the strategy has access to task boundaries, adapt the model
             # for the whole origin experience to add the
             if self.experience.access_task_boundaries:
-                avalanche_model_adaptation(model,
+                avalanche_model_adaptation(model, 
                                            self.experience.origin_experience)
             else:
                 avalanche_model_adaptation(model, self.experience)
-
-        # For evaluation, the experience is not necessarily an online
-        # experience:
         else:
             avalanche_model_adaptation(model, self.experience)
 
         return model.to(self.device)
 
-    def make_optimizer(self):
+    def make_optimizer(self, reset_opt=True, reset_state=False):
         """Optimizer initialization.
 
         Called before each training experiene to configure the optimizer.
         """
-        if isinstance(self.experience, OnlineCLExperience):
-            if self.experience.access_task_boundaries:
-                assert self.experience.is_first_subexp
-                reset_optimizer(self.optimizer, self.model)
+        reset_opt = detect_param_change(
+            self.optimizer, list(self.model.named_parameters()),
+        ) or reset_opt
 
-            # Otherwise, update the optimizer
-            else:
-                update_optimizer(
-                    self.optimizer,
-                    self.model_params_before_adaptation,  # type: ignore
-                    self.model.parameters(),
-                    reset_state=False)
-        else:
+        if reset_state:
```

Comment:
reset_state and reset_opt are both resetting the state. What's the difference?

Reply:
reset state is resetting only the state, not attributing new parameters to the optimizer, whereas reset_opt is doing both (attributing new parameters to the optimizer and resetting the state)

Reply:
> reset state is resetting only the state, not attributing new parameters to the optimizer

we never want to do this. I think you can remove one of the two options since they are redundants.